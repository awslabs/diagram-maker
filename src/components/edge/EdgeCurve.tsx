import * as classnames from 'classnames';
import * as Preact from 'preact';

import { getInflectionPoint } from 'diagramMaker/service/positionUtils';
import { Position } from 'diagramMaker/state/types';

// @NOTE: Depending on node connector placement
// the bezier curve is drawn differently, so we have two types
export enum EdgeStyle {
  LEFT_RIGHT_BEZIER,
  TOP_BOTTOM_BEZIER,
  STRAIGHT,
  QUADRATIC_BEZIER
}

export interface EdgeCurveProps {
  src: Position;
  dest: Position;
  edgeStyle: EdgeStyle;
  className?: string;
  showArrowhead?: boolean;
}

const getQuadraticBezierPath = (src: Position, dest: Position, className?: string, markerEnd?: string): JSX.Element => {
  const { x: x1, y: y1 } = src;
  const { x: x2, y: y2 } = dest;
  const { x: startingX, y: startingY } = getInflectionPoint(src, dest);
  const classNames = classnames('dm-path', className);

  const pathData = `M${x1},${y1} Q${startingX},${startingY} ${x2},${y2}`;
  return <path className={classNames} d={pathData} marker-end={markerEnd} />;
};

const getLeftRightBezierPath = (src: Position, dest: Position, className?: string, markerEnd?: string): JSX.Element => {
  const { x: x1, y: y1 } = src;
  const { x: x2, y: y2 } = dest;
  const width = x2 - x1;
  const halfWidth = width / 2;
  const classNames = classnames('dm-path', className);

  const pathData = `M${x1},${y1} C${x1 + Math.abs(halfWidth)},${y1} ${x2 - Math.abs(halfWidth)},${y2} ${x2},${y2}`;
  return <path className={classNames} d={pathData} marker-end={markerEnd} />;
};

const getTopBottomBezierPath = (src: Position, dest: Position, className?: string, markerEnd?: string): JSX.Element => {
  const { x: x1, y: y1 } = src;
  const { x: x2, y: y2 } = dest;
  const height = y2 - y1;
  const halfHeight = height / 2;
  const classNames = classnames('dm-path', className);

  const pathData = `M${x1},${y1} C${x1},${y1 + Math.abs(halfHeight)} ${x2},${y2 - Math.abs(halfHeight)} ${x2},${y2}`;
  return <path className={classNames} d={pathData} marker-end={markerEnd} />;
};

const getStraightPath = (src: Position, dest: Position, className?: string, markerEnd?: string): JSX.Element => {
  const classNames = classnames('dm-path', className);
  return <line className={classNames} x1={src.x} y1={src.y} x2={dest.x} y2={dest.y} marker-end={markerEnd} />;
};

const EdgeCurve = (props: Preact.RenderableProps<EdgeCurveProps>): JSX.Element => {
  const { src, dest, edgeStyle, className, showArrowhead } = props;
  let markerEnd;
  if (showArrowhead) {
    markerEnd = 'url(#arrow)';
  }
  switch (edgeStyle) {
    case EdgeStyle.LEFT_RIGHT_BEZIER:
      return getLeftRightBezierPath(src, dest, className, markerEnd);
    case EdgeStyle.TOP_BOTTOM_BEZIER:
      return getTopBottomBezierPath(src, dest, className, markerEnd);
    case EdgeStyle.STRAIGHT:
      return getStraightPath(src, dest, className, markerEnd);
    case EdgeStyle.QUADRATIC_BEZIER:
      return getQuadraticBezierPath(src, dest, className, markerEnd);
  }
};

export default EdgeCurve;
