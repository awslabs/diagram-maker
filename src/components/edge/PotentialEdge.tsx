import * as classnames from 'classnames';
import * as Preact from 'preact';

import { DiagramMakerComponentsType } from 'diagramMaker/service/ui/types';
import { Position } from 'diagramMaker/state/types';

import { default as EdgeCurve, EdgeStyle } from './EdgeCurve';

import './Edge.scss';

export interface PotentialEdgeProps {
  src: Position;
  dest: Position;
  edgeStyle: EdgeStyle;
  className?: string | string[];
  showArrowhead?: boolean;
}

const getCurve = (
    src: Position, dest: Position, edgeStyle: EdgeStyle, className: string, showArrowhead?: boolean
  ) =>
    <EdgeCurve src={src} dest={dest} edgeStyle={edgeStyle} className={className} showArrowhead={showArrowhead} />;

const PotentialEdge = (props: Preact.RenderableProps<PotentialEdgeProps>): JSX.Element => {
  const { src, dest, className, edgeStyle, showArrowhead } = props;

  const classNames = classnames('dm-edge', className);

  const curvePath = getCurve(src, dest, edgeStyle, 'dm-path-inner', showArrowhead);

  return(
    <g
      className={classNames}
      data-type={DiagramMakerComponentsType.POTENTIAL_EDGE}
    >
      {curvePath}
    </g>
  );
};

export default PotentialEdge;
