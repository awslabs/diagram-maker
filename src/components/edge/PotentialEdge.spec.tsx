import * as Preact from 'preact';
import { shallow } from 'preact-render-spy';

import { EdgeStyle } from './EdgeCurve';
import PotentialEdge from './PotentialEdge';

describe('PotentialEdge', () => {
  it('renders the paths using props', () => {
    const edge = shallow(
      <PotentialEdge
        edgeStyle={EdgeStyle.LEFT_RIGHT_BEZIER}
        src={{ x:0, y:100 }}
        dest={{ x:100, y:0 }}
      />
    );

    expect(edge).toMatchSnapshot();
  });

  it('passes show arrowhead', () => {
    const edge = shallow(
      <PotentialEdge
        edgeStyle={EdgeStyle.LEFT_RIGHT_BEZIER}
        src={{ x:0, y:100 }}
        dest={{ x:100, y:0 }}
        showArrowhead={true}
      />
    );

    expect(edge).toMatchSnapshot();
  });
});
