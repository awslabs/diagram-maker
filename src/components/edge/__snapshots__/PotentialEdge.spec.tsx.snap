// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PotentialEdge passes show arrowhead 1`] = `
preact-render-spy (1 nodes)
-------
<g
  class="dm-edge"
  data-type="DiagramMaker.PotentialEdge"
>
  <EdgeCurve
    src={
      Object {
        "x": 0,
        "y": 100
      }
    }
    dest={
      Object {
        "x": 100,
        "y": 0
      }
    }
    edgeStyle={0}
    class="dm-path-inner"
    showArrowhead={true}
  >
  </EdgeCurve>
</g>

`;

exports[`PotentialEdge renders the paths using props 1`] = `
preact-render-spy (1 nodes)
-------
<g
  class="dm-edge"
  data-type="DiagramMaker.PotentialEdge"
>
  <EdgeCurve
    src={
      Object {
        "x": 0,
        "y": 100
      }
    }
    dest={
      Object {
        "x": 100,
        "y": 0
      }
    }
    edgeStyle={0}
    class="dm-path-inner"
  >
  </EdgeCurve>
</g>

`;
